versioned/persist/tx/src/main/java/org/projectnessie/versioned/persist/tx/TxConnectionProvider.java

```diff
@@ -0,0 +1,151 @@
+      boolean metadataUpperCase,
+      boolean batchDDL,
+      String catalog,
+      String schema) {
+    try (Connection c = this.borrowConnection()) {
+      try (Statement st = c.createStatement()) {
+        Object[] formatParamsArray = formatParams.toArray();
+        Stream<String> ddls =
+            perTableDDLs.entrySet().stream()
+                .filter(e -> !tableExists(c, metadataUpperCase, catalog, schema, e.getKey()))
+                .flatMap(e -> e.getValue().stream())
+                .map(s -> MessageFormat.format(s, formatParamsArray));
+
+        executeDDLs(batchDDL, ddls, st);
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void executeDDLs(boolean batchDDL, Stream<String> ddls, Statement st)
+      throws SQLException {
+    if (batchDDL) {
+      String ddl = ddls.map(s -> s + ";").collect(Collectors.joining("\n\n"));
+      if (!ddl.isEmpty()) {
+        ddl = "BEGIN;\n" + ddl + "\nEND TRANSACTION;\n";
+        st.execute(ddl);
+      }
+    } else {
+      List<String> ddlStatements = ddls.collect(Collectors.toList());
+      for (String ddl : ddlStatements) {
+        st.execute(ddl);
+      }
+    }
+  }
+
+  private static boolean tableExists(
+      Connection conn, boolean metadataUpperCase, String catalog, String schema, String table) {
+    if (metadataUpperCase) {
+      catalog = catalog != null ? catalog.toUpperCase(Locale.ROOT) : null;
+      schema = schema != null ? schema.toUpperCase(Locale.ROOT) : null;
+      table = table != null ? table.toUpperCase(Locale.ROOT) : null;
+    }
+
+    try (ResultSet tables = conn.getMetaData().getTables(catalog, schema, table, null)) {
+      return tables.next();
+    } catch (SQLException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void close() throws Exception {}
+
+  public Connection borrowConnection() throws SQLException {
+    throw new UnsupportedOperationException();
+  }
+
+  public static class LocalConnectionProvider extends TxConnectionProvider {
+    private DataSource dataSource;
+
+    @Override
+    public void configure(TxDatabaseAdapterConfig config) {
+      AgroalDataSourceConfigurationSupplier dataSourceConfiguration =
+          new AgroalDataSourceConfigurationSupplier();
+      AgroalConnectionPoolConfigurationSupplier poolConfiguration =
+          dataSourceConfiguration.connectionPoolConfiguration();
+      AgroalConnectionFactoryConfigurationSupplier connectionFactoryConfiguration =
+          poolConfiguration.connectionFactoryConfiguration();
+
+      // configure pool
+      poolConfiguration
+          .initialSize(10)
+          .maxSize(10)
+          .minSize(10)
+          .maxLifetime(Duration.of(5, ChronoUnit.MINUTES))
+          .acquisitionTimeout(Duration.of(30, ChronoUnit.SECONDS));
+
+      // configure supplier
+      connectionFactoryConfiguration.jdbcUrl(config.getJdbcUrl());
+      if (config.getJdbcUser() != null) {
```

Comment:
should it go into test module then?

Reply:
Hm - that would become a cyclic dependency.

Reply:
I am a bit confused. We just added more config for the pool here. This can't be decoupled from teh rest of `main`, it looks (to me w/o going into detail) that its a key class here. From this conversation it sounds like some of it should be reused and the rest put into the `test` module. Thoughts?

Reply:
It's really tricky to extract that to or as a "test only" dependency. I could extract it into a separate package though. so you have to explicitly "ask for" the `LocalConnectionProvider` by using `LocalDatabaseAdapterConfig`.